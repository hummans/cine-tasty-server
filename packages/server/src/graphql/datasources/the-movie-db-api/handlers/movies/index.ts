import {
  BaseMovie,
  TrendingMoviesQueryResult,
  QueryMovieArgs,
  Movie,
  MovieReviewsArgs,
  SimilarMoviesQueryResult,
  ReviewsQueryResult,
  ReviewItem,
  MovieSimilarArgs,
  TrendingMoviesArgs,
} from '../../../../../lib/types';
import { GetTMDBApiRequest, TrendingMoviesEndpoints } from '../../../../../types';

const BASE_ENDPOINT = 'movie';

type BaseResponse = {
  total_results: number;
  total_pages: number;
  page: number;
};

type GetBaseMovieResponse = BaseResponse & {
  results: BaseMovie[];
};

type GetReviewsResponse = BaseResponse & {
  results: ReviewItem[];
};

type GetRequestParams = { page: number } | { append_to_response: string };

export interface Props {
  getSimilars(params: MovieSimilarArgs): Promise<SimilarMoviesQueryResult>;
  getReviews({ id, reviewsPage }: MovieReviewsArgs): Promise<ReviewsQueryResult>;
  getMovie: (params: QueryMovieArgs) => Promise<Movie | null>;
  getTrendingItem: (
    params: TrendingMoviesArgs,
    resource: TrendingMoviesEndpoints,
  ) => Promise<TrendingMoviesQueryResult>;
}

class MovieHandler implements Props {
  get: GetTMDBApiRequest;

  constructor(execGetRequest: GetTMDBApiRequest) {
    this.get = execGetRequest;
  }

  async getMovie({ id, language }: QueryMovieArgs): Promise<Movie | null> {
    const movie = await this.get<
      GetRequestParams,
      Promise<Movie & { status_message?: string }>
    >(`${BASE_ENDPOINT}/${id}`, { append_to_response: 'videos,credits' }, language);

    if (typeof movie.status_message === 'string') {
      return null;
    }

    return movie;
  }

  async getReviews({ id, reviewsPage }: MovieReviewsArgs): Promise<ReviewsQueryResult> {
    const { total_pages: totalPages, total_results, results } = await this.get<
      GetRequestParams,
      Promise<GetReviewsResponse>
    >(`${BASE_ENDPOINT}/${id}/reviews`, {
      page: reviewsPage,
    });

    return {
      hasMore: reviewsPage < totalPages,
      total_pages: totalPages,
      total_results,
      items: results,
    };
  }

  async getSimilars({
    id,
    similarsPage,
    language,
  }: MovieSimilarArgs): Promise<SimilarMoviesQueryResult> {
    const { total_pages: totalPages, total_results, results } = await this.get<
      GetRequestParams,
      Promise<GetBaseMovieResponse>
    >(
      `${BASE_ENDPOINT}/${id}/similar`,
      {
        page: similarsPage,
      },
      language,
    );

    return {
      hasMore: similarsPage < totalPages,
      total_pages: totalPages,
      total_results,
      items: results,
    };
  }

  async getTrendingItem(
    { page, language }: TrendingMoviesArgs,
    resource: TrendingMoviesEndpoints,
  ): Promise<TrendingMoviesQueryResult> {
    const { total_pages: totalPages, total_results, results } = await this.get<
      GetRequestParams,
      Promise<GetBaseMovieResponse>
    >(resource, { page }, language);

    return {
      hasMore: page < totalPages,
      total_pages: totalPages,
      total_results,
      items: results,
    };
  }
}

export default MovieHandler;
