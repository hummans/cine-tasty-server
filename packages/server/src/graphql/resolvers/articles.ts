import { QueryResolvers } from '../../lib/types';
import { Datasource } from '../datasources';
import { Article } from '../../lib/types';

const resolvers: QueryResolvers = {
  Query: {
    articles: (
      _: {},
      __: {},
      { dataSources }: Datasource,
    ): Promise<Article[]> => dataSources.news.getAllArticles(),
  },
};

export default resolvers;
