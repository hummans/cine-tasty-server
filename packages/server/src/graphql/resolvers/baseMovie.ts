import MediaGenresHandler from '../datasources/the-movie-db-api/handlers/media-genres';
import {
  BaseMovieGenre_IdsArgs as BaseMovieGenreIdsArgs,
  QueryResolvers,
  MediaType,
} from '../../lib/types';
import { MediaItem } from '../../types';

const mediaGenres = new MediaGenresHandler();

const resolvers: QueryResolvers = {
  BaseMovie: {
    genre_ids: (
      { genre_ids }: MediaItem,
      { language }: BaseMovieGenreIdsArgs,
    ): Promise<string[]> =>
      mediaGenres.getMediaGenres(genre_ids, MediaType.Movie.toLowerCase(), language),
  },
};

export default resolvers;
