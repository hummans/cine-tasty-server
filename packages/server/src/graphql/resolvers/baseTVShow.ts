import MediaGenresHandler from '../datasources/the-movie-db-api/handlers/media-genres';
import {
  BaseTvShowGenre_IdsArgs as BaseTvShowGenreIdsArgs,
  QueryResolvers,
  MediaType,
} from '../../lib/types';
import { MediaItem } from '../../types';

const mediaGenres = new MediaGenresHandler();

const resolvers: QueryResolvers = {
  BaseTVShow: {
    genre_ids: (
      { genre_ids }: MediaItem,
      { language }: BaseTvShowGenreIdsArgs,
    ): Promise<string[]> =>
      mediaGenres.getMediaGenres(genre_ids, MediaType.Tv.toLowerCase(), language),
  },
};

export default resolvers;
