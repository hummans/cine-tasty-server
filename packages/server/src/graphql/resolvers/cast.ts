import { QueryResolvers, MediaType } from '../../lib/types';

type CastType = {
  media_type: 'tv' | 'movie';
};

const resolveCastType = (cast: CastType): string | null => {
  if (cast.media_type === MediaType.Movie.toLowerCase()) {
    return 'CastMovie';
  }

  if (cast.media_type === MediaType.Tv.toLowerCase()) {
    return 'CastTV';
  }

  return null;
};

const resolveTypes: QueryResolvers = {
  Cast: {
    __resolveType(cast: CastType): string | null {
      return resolveCastType(cast);
    },
  },
};

export default resolveTypes;
