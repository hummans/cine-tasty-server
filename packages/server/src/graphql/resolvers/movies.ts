import MediaGenresHandler from '../datasources/the-movie-db-api/handlers/media-genres';

import {
  QueryResolvers,
  TrendingMoviesQueryResult,
  TrendingMoviesNow_PlayingArgs as TrendingMoviesNowPlayingArgs,
  TrendingMoviesPopularArgs,
  TrendingMoviesTop_RatedArgs as TrendingMoviesTopRatedArgs,
  TrendingMoviesUpcomingArgs,
  QueryMovieArgs,
  BaseMovieGenre_IdsArgs as BaseMovieGenreIdsArgs,
  Movie,
  MovieGenresArgs,
  MediaType,
  CastItem,
  CrewItem,
  MediaVideo,
  MovieReviewsArgs,
  ReviewsQueryResultResolvers,
  SimilarMoviesQueryResultResolvers,
  MovieSimilarArgs,
} from '../../lib/types';
import { Context, MediaGenre, MediaItem, MediaCredits } from '../../types';

const mediaGenres = new MediaGenresHandler();

type MovieVideos = {
  results: MediaVideo[];
};

const BASE_VIDEO_THHUMBNAIL_URL = 'https://img.youtube.com/vi';

const resolvers: QueryResolvers = {
  Query: {
    trending_movies: (): {} => ({}),

    movie: (
      _: {},
      args: QueryMovieArgs,
      { dataSources }: Context,
    ): Promise<Movie | null> => dataSources.tmdb.getMovie(args),
  },

  BaseMovie: {
    genre_ids: (
      { genre_ids }: MediaItem,
      { language }: BaseMovieGenreIdsArgs,
    ): Promise<string[]> =>
      mediaGenres.getMediaGenres(genre_ids, MediaType.Movie.toLowerCase(), language),
  },

  Movie: {
    spoken_languages: ({
      spoken_languages,
    }: {
      spoken_languages: Array<{ name: string }>;
    }): string[] => spoken_languages.map(({ name }) => name),

    production_countries: ({
      production_countries,
    }: {
      production_countries: Array<{ name: string }>;
    }): string[] => production_countries.map(({ name }) => name),

    cast: ({ credits }: { credits: MediaCredits }): CastItem[] =>
      credits.cast.map(castItem => ({
        name: castItem.name,
        profile_path: castItem.profile_path,
        id: castItem.id,
        character: castItem.character,
      })),

    crew: ({ credits }: { credits: MediaCredits }): CrewItem[] =>
      credits.crew.map(castItem => ({
        profile_path: castItem.profile_path,
        department: castItem.department,
        id: castItem.id,
        job: castItem.job,
        name: castItem.name,
      })),

    reviews: (
      _: {},
      args: MovieReviewsArgs,
      { dataSources }: Context,
    ): ReviewsQueryResultResolvers => dataSources.tmdb.getMovieReviews(args),

    similar: (
      _: {},
      args: MovieSimilarArgs,
      { dataSources }: Context,
    ): SimilarMoviesQueryResultResolvers => dataSources.tmdb.getSimilarMovies(args),

    videos: ({ videos }: { videos: MovieVideos }): MediaVideo[] =>
      videos.results
        .filter(({ site }) => site === 'YouTube')
        .map(video => ({
          key: video.key,
          name: video.name,
          site: video.site,
          id: `${video.id}`,
          type: video.type,
          thumbnail: {
            extra_small: `${BASE_VIDEO_THHUMBNAIL_URL}/${video.key}/default.jpg`,
            small: `${BASE_VIDEO_THHUMBNAIL_URL}/${video.key}/mqdefault.jpg`,
            medium: `${BASE_VIDEO_THHUMBNAIL_URL}/${video.key}/hqdefault.jpg`,
            large: `${BASE_VIDEO_THHUMBNAIL_URL}/${video.key}/sddefault.jpg`,
            extra_large: `${BASE_VIDEO_THHUMBNAIL_URL}/${video.key}/maxresdefault.jpg`,
          },
        })),

    genres: (
      { genres }: { genres: Array<MediaGenre> },
      { language }: MovieGenresArgs,
    ): Promise<string[]> => {
      const genreIds = genres.map(({ id }) => id);

      return mediaGenres.getMediaGenres(
        genreIds,
        MediaType.Movie.toLowerCase(),
        language,
      );
    },
  },

  TrendingMovies: {
    now_playing: (
      _: {},
      { args }: TrendingMoviesNowPlayingArgs,
      { dataSources }: Context,
    ): Promise<TrendingMoviesQueryResult> =>
      dataSources.tmdb.getTrendingMoviesItem(args, 'now_playing'),

    popular: (
      _: {},
      { args }: TrendingMoviesPopularArgs,
      { dataSources }: Context,
    ): Promise<TrendingMoviesQueryResult> =>
      dataSources.tmdb.getTrendingMoviesItem(args, 'popular'),

    top_rated: (
      _: {},
      { args }: TrendingMoviesTopRatedArgs,
      { dataSources }: Context,
    ): Promise<TrendingMoviesQueryResult> =>
      dataSources.tmdb.getTrendingMoviesItem(args, 'top_rated'),

    upcoming: (
      _: {},
      { args }: TrendingMoviesUpcomingArgs,
      { dataSources }: Context,
    ): Promise<TrendingMoviesQueryResult> =>
      dataSources.tmdb.getTrendingMoviesItem(args, 'upcoming'),
  },
};

export default resolvers;
