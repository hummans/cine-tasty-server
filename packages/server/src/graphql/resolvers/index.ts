import articles from './articles';
import baseMovie from './baseMovie';
import baseTVShow from './baseTVShow';
import cast from './cast';
import castMovie from './castMovie';
import castTV from './castTV';
import people from './people';
import person from './person';
import personKnownFor from './personKnownFor';
import search from './search';
import searchResultItem from './searchResultItem';

export default [
  articles,
  baseMovie,
  baseTVShow,
  cast,
  castTV,
  castMovie,
  people,
  person,
  personKnownFor,
  search,
  searchResultItem,
];
