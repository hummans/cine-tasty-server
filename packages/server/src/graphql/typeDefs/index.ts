import ISO6391Language from './types/ISO6391Language';
import root from './root';
import baseMovie from './types/baseMovie';
import basePerson from './types/basePerson';
import baseTVShow from './types/baseTVShow';
import cast from './types/cast';
import castMovie from './types/castMovie';
import castTV from './types/castTVShow';
import articleLanguage from './types/articleLanguage';
import mediaType from './types/mediaType';
import article from './types/article';
import articleQueryResult from './types/articleQueryResult';
import peopleQueryResult from './types/peopleQueryResult';
import person from './types/person';
import searchResults from './types/searchResults';
import searchTypes from './types/searchTypes';
import trendingMovies from './types/trendingMovies';
import queries from './queries';

export default [
  root,
  trendingMovies,
  article,
  peopleQueryResult,
  searchTypes,
  articleQueryResult,
  articleLanguage,
  baseMovie,
  basePerson,
  baseTVShow,
  searchResults,
  cast,
  person,
  mediaType,
  castMovie,
  castTV,
  person,
  ISO6391Language,
  queries,
];
