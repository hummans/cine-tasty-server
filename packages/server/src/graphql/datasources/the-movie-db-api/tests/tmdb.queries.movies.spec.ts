import { createTestClient } from 'apollo-server-testing';
import { ApolloServer, gql } from 'apollo-server';

import { movieGenres } from '../../../../__tests__/mocks/mediaGenres.stub';
import { rawMovie, movie } from '../../../../__tests__/mocks/movies.stub';

const mockRestDataSourceGet = jest.fn();

import env from '../../../../config/environment';
import resolvers from '../../../resolvers';
import typeDefs from '../../../typeDefs';
import TheMovieDBAPI from '..';

const GENRE_MOVIE_ENDPOINT = '/genre/movie/list';
const NOW_PLAYING_ENDPOINT = 'movie/now_playing';

const GET_TRENDING_MOVIES = gql`
  fragment TrendingMovieItem on BaseMovie {
    genre_ids
    original_title
    video
    title
    adult
    release_date
    backdrop_path
    overview
    vote_average
    poster_path
    popularity
    original_language
    vote_count
    overview
    id
  }

  query TrendingMovies($page: Int!) {
    trending_movies {
      now_playing(input: { page: $page }) {
        total_results
        total_pages
        hasMore
        items {
          ...TrendingMovieItem
        }
      }
      popular(input: { page: $page }) {
        total_results
        total_pages
        hasMore
        items {
          ...TrendingMovieItem
        }
      }
    }
  }
`;

const makeTestServer = (): ApolloServer => {
  return new ApolloServer({
    typeDefs,
    resolvers,
    dataSources: () => ({
      tmdb: new TheMovieDBAPI(),
    }),
  });
};

jest.mock('apollo-datasource-rest', () => {
  class MockRESTDataSource {
    baseUrl = '';
    get = mockRestDataSourceGet;
  }

  return {
    RESTDataSource: MockRESTDataSource,
    HTTPCache: class HTTPCache {},
  };
});

describe('[TheMovieDBAPI.Queries.Movies]', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('fetches the now playing/popular movies from TheMovieDB API and returns the result correctly', async () => {
    mockRestDataSourceGet
      .mockReturnValueOnce({
        total_pages: 1,
        total_results: 1,
        results: [rawMovie],
      })
      .mockReturnValueOnce({
        total_pages: 1,
        total_results: 1,
        results: [rawMovie],
      })
      .mockReturnValueOnce({ genres: movieGenres })
      .mockReturnValueOnce({ genres: movieGenres });

    const server = makeTestServer();

    const { query } = createTestClient(server);

    const { data } = await query({
      query: GET_TRENDING_MOVIES,
      variables: { page: 1 },
    });

    expect(mockRestDataSourceGet.mock.calls.length).toBe(4);

    expect(mockRestDataSourceGet).toHaveBeenCalledWith(NOW_PLAYING_ENDPOINT, {
      api_key: env.THE_MOVIE_DB_API_KEY,
      language: 'en-us',
      page: 1,
    });

    expect(mockRestDataSourceGet).toHaveBeenCalledWith(GENRE_MOVIE_ENDPOINT, {
      api_key: env.THE_MOVIE_DB_API_KEY,
      language: 'en-us',
    });

    expect(data!.trending_movies).toEqual({
      now_playing: {
        hasMore: false,
        total_pages: 1,
        total_results: 1,
        items: [movie],
      },
      popular: {
        hasMore: false,
        total_pages: 1,
        total_results: 1,
        items: [movie],
      },
    });
  });
});
