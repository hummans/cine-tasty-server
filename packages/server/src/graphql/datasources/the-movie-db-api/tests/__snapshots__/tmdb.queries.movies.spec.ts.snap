// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration: DataSources-Movies Query - Movie Detail fetches the now playing/popular/top_rated/upcoming movies from TheMovieDB API and returns the result correctly 1`] = `
Object {
  "now_playing": Object {
    "hasMore": false,
    "items": Array [
      Object {
        "adult": false,
        "backdrop_path": "/backdrop_path.jpg",
        "genre_ids": Array [
          "Movie Genre",
        ],
        "id": 530915,
        "original_language": "original_language",
        "original_title": "original_title",
        "overview": "overview",
        "popularity": 222.017,
        "poster_path": "/poster_path.jpg",
        "release_date": "release_date",
        "title": "title",
        "video": false,
        "vote_average": 8.1,
        "vote_count": 1530,
      },
    ],
    "total_pages": 1,
    "total_results": 1,
  },
  "popular": Object {
    "hasMore": false,
    "items": Array [
      Object {
        "adult": false,
        "backdrop_path": "/backdrop_path.jpg",
        "genre_ids": Array [
          "Movie Genre",
        ],
        "id": 530915,
        "original_language": "original_language",
        "original_title": "original_title",
        "overview": "overview",
        "popularity": 222.017,
        "poster_path": "/poster_path.jpg",
        "release_date": "release_date",
        "title": "title",
        "video": false,
        "vote_average": 8.1,
        "vote_count": 1530,
      },
    ],
    "total_pages": 1,
    "total_results": 1,
  },
  "top_rated": Object {
    "hasMore": false,
    "items": Array [
      Object {
        "adult": false,
        "backdrop_path": "/backdrop_path.jpg",
        "genre_ids": Array [
          "Movie Genre",
        ],
        "id": 530915,
        "original_language": "original_language",
        "original_title": "original_title",
        "overview": "overview",
        "popularity": 222.017,
        "poster_path": "/poster_path.jpg",
        "release_date": "release_date",
        "title": "title",
        "video": false,
        "vote_average": 8.1,
        "vote_count": 1530,
      },
    ],
    "total_pages": 1,
    "total_results": 1,
  },
  "upcoming": Object {
    "hasMore": false,
    "items": Array [
      Object {
        "adult": false,
        "backdrop_path": "/backdrop_path.jpg",
        "genre_ids": Array [
          "Movie Genre",
        ],
        "id": 530915,
        "original_language": "original_language",
        "original_title": "original_title",
        "overview": "overview",
        "popularity": 222.017,
        "poster_path": "/poster_path.jpg",
        "release_date": "release_date",
        "title": "title",
        "video": false,
        "vote_average": 8.1,
        "vote_count": 1530,
      },
    ],
    "total_pages": 1,
    "total_results": 1,
  },
}
`;

exports[`Integration: DataSources-Movies Query - Movie Reviews fetches the reviews of a movie from TheMovieDB API and returns the result correctly and returns hasMore field as true when has more items to be pagianted 1`] = `
Object {
  "hasMore": true,
  "items": Array [
    Object {
      "author": "author.",
      "content": "content",
      "id": "id",
      "url": "url",
    },
  ],
  "total_pages": 2,
  "total_results": 2,
}
`;

exports[`Integration: DataSources-Movies Query - Movie Reviews should query the reviews of a movie from TheMovieDB API and returns the result correctly 1`] = `
Object {
  "hasMore": false,
  "items": Array [
    Object {
      "author": "author.",
      "content": "content",
      "id": "id",
      "url": "url",
    },
  ],
  "total_pages": 1,
  "total_results": 1,
}
`;
